### ISSUE TITLE
bug: Issues with multibyte characters on textbox fields

### ISSUE THREAD

**Issue by sergio-ropero** [openemr/openemr#6706]
### Describe the bug

Inserting a text that fits the field size gives an error when using multibyte characters.
It's pretty common using UTF8 to have these multibyte characters (example: `é`, `ä`, ...) and although they are allowed in MariaDB database, the `LengthBetween` validation used in OpenEMR block them to reach the DB.

### To Reproduce
Steps to reproduce the behaviour:

1. Search for a patient.
2. Click on edit `Demographics` for the patient:
<img width="765" alt="Edit" src="https://github.com/openemr/openemr/assets/42538006/8dba4541-86d0-437e-832e-7a7c602c23f3">

3. Add `ééééééé` on S.S field (7 character text):
<img width="1101" alt="Fill" src="https://github.com/openemr/openemr/assets/42538006/d2e034df-a65d-4acb-a905-1a0dee8e49f2">

4. Click on save

6. See error:
<img width="209" alt="Error" src="https://github.com/openemr/openemr/assets/42538006/d9e74517-9f7f-421e-9e39-9e4bb21ea0c3">

### Expected behaviour
The data is persisted in the database.

As example, I created a table in `mariadb` with the same definition and inserted the sample data in the `To reproduce` section. As we can see, the database can handle the input (although the size is 14 due to multibyte characters):
<img width="427" alt="DB insert" src="https://github.com/openemr/openemr/assets/42538006/5a1c0f42-e8d4-49bf-846a-becd997c1971">
Script:
```
CREATE TABLE issue (txt VARCHAR(11));
INSERT INTO issue VALUES ('ééééééé');
SELECT txt, LENGTH(txt) FROM issue;
```

Sadly the validation [LengthBetween](https://github.com/particle-php/Validator/blob/aab07e734bc40e8777fb4a1f9d9037dc11fb60b6/src/Rule/LengthBetween.php#L70) (which uses `strlen` under the hood) doesn't allow to persist the data. Instead `mb_strlen` can handle UTF-8 multibyte characters as expected.

### Client configuration

- Browser: Chrome 115.0.5790.114
- OpenEMR version: 7.0.0
- Operating system: MacOS Ventura 13.5

**Comment by stephenwaite** [comment]
hi @sergio-ropero , thanks for the report, fixed the underling `strlen` call.

**Comment by sergio-ropero** [comment]
Hello @stephenwaite,

The solution you proposed only fix the input from the form. For the services, I see that OpenEMR is using an external library [Validator](https://github.com/particle-php/Validator) (see [here](https://github.com/openemr/openemr/blob/e856d651d868aa895a8b336d269a88b5962dae70/src/Services/EncounterService.php#L33) as example).

Also I realise that there is an [old bug](https://github.com/particle-php/Validator/issues/97#issuecomment-1659703037) on that library for the multibyte issue. I have ping them but the issue is from 2016.

[These are all the places](https://github.com/search?q=repo%3Aopenemr%2Fopenemr%20Particle%5CValidator%5CValidator&type=code) in which OpenEMR is using the library and can be impacted by the bug.

What I was able to verify that it only impacts to `Lenght` and `LengthBetween`.

**Comment by stephenwaite** [comment]
Hi @sergio-ropero, can you post an error here from OpenEMR that you think is due to particle's validator? We can hopefully make sure there's plenty of room in the db for these fields that maybe impacted. Thank you.

**Comment by stephenwaite** [comment]
okay, thanks @sergio-ropero , looks like [this PR](https://github.com/particle-php/Validator/pull/105/files) could've fixed the issue but was closed.

**Comment by sergio-ropero** [comment]
Indeed, but checking the discussion there were some points of failure (like the `mb_detect_encoding` does not return an encoding).

**Comment by stephenwaite** [comment]
Okay, it looks like there needs to be a new PR. Maybe that is something that you could submit?

**Comment by sergio-ropero** [comment]
Hello Stephen,
Do you know how much effort this change could take on your side?

FYI: @rehmanatcardo

**Comment by stephenwaite** [comment]
hi @sergio-ropero , be best to fix it upstream.

**Comment by adunsulag** [comment]
@sergio-ropero I pinged the upstream project maintainer here on this closed PR https://github.com/particle-php/Validator/pull/161 which looks like once that is upstream we can add into our validators the correct UTF8 encoding.   The other option would be to go in and write a custom validator to do the mb_strlen check, not hard, but preference would be to get the fix in upstream.

**Comment by bradymiller** [comment]
If can not fix in upstream, then another option would be to fork the repo into openemr org and then update it there, and then bring our fork via composer instead of the original package.

**Comment by rehmanatcardo** [comment]
@adunsulag , @stephenwaite ,  can you please help close this ticket.

**Comment by stephenwaite** [comment]
hi @rehmanatcardo , why would you like it closed? Were you able to find a solution? thank you

**Comment by stephenwaite** [comment]
okay, it looks like @bradymiller has the midas touch in that repo with the latest PR :)
![Screenshot from 2023-08-09 11-46-05](https://github.com/openemr/openemr/assets/5618372/55ccc5f9-3002-4610-a705-fa9e336b7121)

https://github.com/openemr/openemr/assets/5618372/5f69e66d-fe50-495e-867c-dfe86cce6a34

**Comment by adunsulag** [comment]
It does not look like the Validator project is really responding at this point.  We will need to either fork the project or write a custom validator to handle multibyte characters.  This would be a high impact change and I don't want to bring it in this week for release 7.0.2 so marking this issue to be for 7.0.2.1
